apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtension.java

```diff
@@ -0,0 +1,325 @@
+   private final NeutronSecurityGroupToSecurityGroup.Factory neutronSecurityGroupToSecurityGroup;
+
+   @Inject(optional = true)
+   @Named("openstack-neutron")
+   private Supplier<Context> neutronContextSupplier;
+
+   @Inject
+   NeutronSecurityGroupExtension(NovaApi api,
+                                 @Region Supplier<Set<String>> regionIds,
+                                 GroupNamingConvention.Factory namingConvention, LoadingCache<RegionAndName, SecurityGroup> groupCreator,
+                                 Supplier<Map<String, Location>> locationIndex,
+                                 NeutronSecurityGroupToSecurityGroup.Factory neutronSecurityGroupToSecurityGroup) {
+      this.api = api;
+      this.regionIds = checkNotNull(regionIds, "regionIds");
+      this.namingConvention = checkNotNull(namingConvention, "namingConvention");
+      this.groupCreator = groupCreator;
+      this.locationIndex = locationIndex;
+      this.neutronSecurityGroupToSecurityGroup = neutronSecurityGroupToSecurityGroup;
+   }
+
+   @Override
+   public SecurityGroup createSecurityGroup(String name, Location location) {
+      String region = location.getId();
+      if (region == null) {
+         return null;
+      }
+      logger.debug(">> creating security group %s in %s...", name, location);
+
+      String markerGroup = namingConvention.create().sharedNameForGroup(name);
+      RegionSecurityGroupNameAndPorts regionAndName = new RegionSecurityGroupNameAndPorts(region, markerGroup, ImmutableSet.<Integer> of());
+      return groupCreator.getUnchecked(regionAndName);
+   }
+
+   @Override
+   public Set<SecurityGroup> listSecurityGroups() {
+      Set<SecurityGroup> securityGroups = Sets.newHashSet();
+
+      for (String regionId : regionIds.get()) {
+         Location location = locationIndex.get().get(regionId);
+         securityGroups.addAll(listSecurityGroupsInLocation(location));
+      }
+      return ImmutableSet.copyOf(securityGroups);
+   }
+
+
+   @Override
+   public Set<SecurityGroup> listSecurityGroupsInLocation(final Location location) {
+      String region = location.getId();
+      if (region == null) {
+         return ImmutableSet.of();
+      }
+      return getSecurityGroupApi(region).listSecurityGroups().concat().transform(neutronSecurityGroupToSecurityGroup.create(location)).toSet();
+   }
+
+   @Override
+   public Set<SecurityGroup> listSecurityGroupsForNode(String id) {
+      RegionAndId regionAndId = RegionAndId.fromSlashEncoded(checkNotNull(id, "id"));
+      String region = regionAndId.getRegion();
+
+      SecurityGroupApi securityGroupApi = getSecurityGroupApi(region);
+
+      final FluentIterable<org.jclouds.openstack.neutron.v2.domain.SecurityGroup> allGroups = securityGroupApi.listSecurityGroups().concat();
+      Location location = locationIndex.get().get(region);
+      return ImmutableSet.copyOf(transform(filter(allGroups, notNull()), neutronSecurityGroupToSecurityGroup.create(location)));
+   }
```

Comment:
This implementation is still wrong. You have to filter ONLY the security groups associated with the given node. You're returning all existing ones.

Reply:
good point! thanks

unfortunately also `NovaSecurityGroupExtension` is not really reliable as it is built on
`Optional<? extends ServerWithSecurityGroupsApi> serverWithSecurityGroupsApi = api.getServerWithSecurityGroupsApi(region)` which is most of the time not working on recent Openstack installations (they claim the extension is there but it is not working!)

We may need to radically change the implementation but not sure how yet

Reply:
I think we should use https://developer.openstack.org/api-ref/compute/#servers-security-groups-servers-os-security-groups

Reply:
+1 to use this. I'd say we need it before merging the PR. Agree?